src/Features/CSharp/Portable/ImplementInterface/AbstractChangeImplementionCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,257 @@
+        }
+
+        private int TotalCount(MemberImplementationMap dictionary)
+        {
+            var result = 0;
+            foreach (var (key, values) in dictionary)
+            {
+                result += values.Count;
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// Returns a mapping from members in our containing types to all the interface members (of
+        /// the sort we care about) that it implements.
+        /// </summary>
+        private MemberImplementationMap GetImplementedMembers(INamedTypeSymbol containingType, ImmutableArray<INamedTypeSymbol> interfaceTypes)
+        {
+            var result = new MemberImplementationMap();
+            foreach (var interfaceType in interfaceTypes)
+            {
+                foreach (var interfaceMember in interfaceType.GetMembers())
+                {
+                    var impl = containingType.FindImplementationForInterfaceMember(interfaceMember);
+                    if (impl != null &&
+                        containingType.Equals(impl.ContainingType) &&
+                        CheckMemberCanBeConverted(impl) &&
+                        !impl.IsAccessor())
+                    {
+                        result.Add(impl, interfaceMember);
+                    }
+                }
+            }
+
+            return result;
+        }
+
+        private async Task<Solution> ChangeImplementationAsync(
+            Project project, MemberImplementationMap implMemberToInterfaceMembers, CancellationToken cancellationToken)
+        {
+            var solution = project.Solution;
+            var solutionEditor = new SolutionEditor(solution);
+
+            // First, we have to go through and find all the references to these interface
+            // implementation members.  We may have to update them to preserve semantics.  i.e. a
+            // call to goo.Bar() will be updated to `((IGoo)goo).Bar()` if we're switching to
+            // explicit implementation.
+            foreach (var (implMember, interfaceMembers) in implMemberToInterfaceMembers)
+            {
+                await UpdateReferencesAsync(
+                    project, solutionEditor, implMember,
+                    interfaceMembers.First().ContainingType,
+                    cancellationToken).ConfigureAwait(false);
+            }
+
+            // Now, bucket all the implemented members by which document they appear in.
+            // That way, we can update all the members in a specific document in bulk.
+            var documentToImplDeclarations = new OrderedMultiDictionary<Document, (SyntaxNode, SetWithInsertionOrder<ISymbol>)>();
+            foreach (var (implMember, interfaceMembers) in implMemberToInterfaceMembers)
+            {
+                foreach (var syntaxRef in implMember.DeclaringSyntaxReferences)
+                {
+                    var doc = solution.GetDocument(syntaxRef.SyntaxTree);
+                    if (doc != null)
```

Comment:
A debug assert for doc being null might be wise -- I don't mind the release-build error checking but if this returns null it means somebody screwed up somewhere.

Reply:
true.

Reply:
done.